packages/hash/api/src/graphql/resolvers/knowledge/page/update-page-contents.ts

```diff
@@ -0,0 +1,443 @@
+      if (draft.existingEntity) {
+        draft.existingEntity.entityId = placeholderResults.get(
+          draft.existingEntity.entityId,
+        );
+      }
+      if (draft.entityType?.entityTypeId) {
+        draft.entityType.entityTypeId = placeholderResults.get(
+          draft.entityType.entityTypeId,
+        );
+      }
+
+      /**
+       * @todo Figure out what would be the equivalent to linked data in the new graph api.
+       *   Related to https://app.asana.com/0/1200211978612931/1201850801682936/f
+       *   Asana ticket: https://app.asana.com/0/1202805690238892/1203045933021781/f
+       */
+      // if (draft.entityProperties?.text?.__linkedData?.entityId) {
+      //   draft.entityProperties.text.__linkedData.entityId =
+      //     placeholderResults.get(
+      //       draft.entityProperties.text.__linkedData.entityId,
+      //     );
+      // }
+    });
+
+    return await EntityModel.createEntityWithLinks(graphApi, {
+      createdById: entityAccountId,
+      entityDefinition,
+    });
+  };
+
+  /**
+   * @todo Figure out how we want to implement entity type creation
+   *   in update-page-contents
+   *   see https://app.asana.com/0/1202805690238892/1203057486837598/f
+   */
+  // Create any _new_ entity types
+  filterForAction(actions, "createEntityType").map(({ index }) => {
+    throw new Error(
+      `createEntityType: not implemented yet, action index: ${index}`,
+    );
+  });
+
+  /**
+   * Create any _new_ entities. This is done one at a time in order to allow
+   * you to reference a previous created entity using its placeholder.
+   */
+  for (const { action, index } of filterForAction(actions, "createEntity")) {
+    await createNewEntity({
+      createEntityAction: action,
+      index,
+      placeholderResults,
+      createEntityWithPlaceholders,
+    });
+  }
+
+  // Create any _new_ blocks
+  const insertedBlocks = await Promise.all(
+    filterForAction(actions, "insertBlock").map(({ action, index }) =>
+      insertNewBlock(graphApi, {
+        userModel,
+        insertBlockAction: action,
+        index,
+        createEntityWithPlaceholders,
+        placeholderResults,
+      }),
+    ),
+  );
+
+  // Perform any block data swapping updates.
+  await Promise.all(
+    filterForAction(actions, "swapBlockData").map(({ action }) =>
+      swapBlockData(graphApi, {
+        userModel,
+        swapBlockDataAction: action,
+      }),
+    ),
+  );
+
+  // Perform any entity updates.
+  await Promise.all(
+    filterForAction(actions, "updateEntity").map(async ({ action }) =>
+      updateEntity(graphApi, { userModel, action, placeholderResults }),
+    ),
+  );
+
+  // @todo, perhaps check this exists first?
+  const pageModel = await PageModel.getPageById(graphApi, {
+    entityId: pageEntityId,
+  });
+
+  if (!pageModel) {
+    const msg = `Page with fixed ID ${pageEntityId} not found in account ${accountId}`;
+    throw new ApolloError(msg, "NOT_FOUND");
+  }
```

Comment:
This todo was actually meant for me to consider before opening for review, my bad. But we might as well discuss it here. We did also briefly chat about this off-line.

The comment was about the page potentially not even existing (incorrect `pageEntityId` on mutation call) resulting in all of the above operations making the state of the entities invalid. What we should/could do since we can't rely on transactions is to check if the page exists in the beginning of this resolver's body, and then perhaps refetch the page at the end as well

Reply:
> What we should/could do since we can't rely on transactions is to check if the page exists in the beginning of this resolver's body, and then perhaps refetch the page at the end as well

I'd be okay with this, but not sure how often we expect this method to be called with a non-existent entity. If it is something that we need to handle gracefully, then moving the check to the top of the transaction and adding a todo to remove it once we have transactions is probably the way to go.

I think we should try and avoid investing too much time in writing temporary code like this though, as we know we will need transactions in the near(ish) future.

cc @Alfred-Mountfield

Reply:
> I think we should try and avoid investing too much time in writing temporary code like this though, as we know we will need transactions in the near(ish) future.

Agreed, I wonder how easy it would be to find some hacky stopgap solution for transactions. I'm trying to think of some but I don't really like any of my ideas at the moment.

Reply:
I would be okay with that

Reply:
These thoughts should be captured by 829c3cd25645d471e21716bc27c64591ba6e9347